REST API + extras

TMiguelT opened the pull request:
# Changes
* Created a best-practise REST API, using Flask-RESTful and marshmallow-jsonapi
    * The API conforms to the [JSON API](https://jsonapi.org/format) spec, meaning conformant clients and servers will automatically understand it!
    * The API also supports a basic CSV export function
    * Code mostly located in `megaqc/rest_api/`
    * The API can be accessed at the URL `/restapi/v1`
    * It uses resources, which generally correspond to database Models, so you can `GET /users` to return a list of users, `POST /reports` to create a report, `DELETE /samples/1` to delete sample 1 etc
    * Fully tested! Pytest coverage for the all the new URLs
    * All database models now have an API endpoint, mashmallow schemas and test factories
    * Closes #85 
    * Plot data is now provided as a resource, under `/plots/trends/series`. This can be used by the frontend to get plot data, but I believe this also fixes #71 and fixes #27. Currently this will only export in JSON format, but CSV export should be easy to implement if needed
* Rewrote the query interpreter
    * The code that generates SQL queries from the MegaQC query language
    * Mostly located in `megaqc/rest_api/filters.py`
    * Also comprehensively tested, with a test case for "and", "or", and every query operator
    * Once we get the plots to use this API/function, this will close #48 and will close #70
* Fixed the original MegaQC tests
    * Things like the models, integration and form tests
    * These probably need more tests, but at least they pass now
* Created Marshmallow schemas for all models
    * Mostly located in `megaqc/rest_api/schemas.py`
    * See https://marshmallow.readthedocs.io/en/3.0/
    * Basically, these define how the models are transformed into/from JSON, in the REST API
* Created many more model factories
    * We had one or two of these before, but I've added many more, covering most models in the database
    * Mostly located in `tests/factories.py` 
* Schema changes
    * Small tweaks to the DB schema, mostly adding relationships for easier traversal of models, and adding `ON DELETE CASCADE` to others, so that we don't have to manually delete all associated models when we delete something like a report
    * Located in `megaqc/model/models.py`

ewels commented:
Wow! Sounds brilliant :) I’m away for three weeks now on holiday, but will be back at work 50% at the end of September.. Looking forward to getting stuck into MegaQC again! Let me know if you’d like to merge stuff in before then or anything.

TMiguelT commented:
Yep, so I created this PR so that you have some time to review and merge this before I put in my React frontend PR, which will depend on this API. The earlier you can merge this the better, but no need to rush since my work doesn't depend on this getting merged.

TMiguelT commented:
But let me get the build working before you bother to look at the code, since it should help you reason with how the code works and trust that it does work.

TMiguelT commented:
Tests all fixed!

tsnowlan commented:
This looks awesome!

TMiguelT commented:
When this gets reviewed, something I'd like guidance on is the permissions for each endpoint. Each endpoint can be:

* Admin only (e.g. deleting a report)
* Logged in user only (e.g. submitting a report)
* Anyone (e.g. listing reports)

However, there might be user-related information that we only want that user and the admin to see, e.g. email address. For example, should `GET /users` should probably not return email addresses for all users, unless you're the admin or owner of that address.

Does anyone have a good idea of which endpoints should be restricted with these mechanisms?

TMiguelT commented:
Although I'm sure no-one has looked at this yet, hold off for a few more days because I realised I need to make a few more changes before it's ready to be merged.

TMiguelT commented:
Okay, I reworked this PR a little so that it uses the [JSON API](https://jsonapi.org/format) standard for communication. It doesn't have any vast difference to before, but by being a standard it provides a convenient reference on how the API should be structured. The other advantage of this is that we can use some existing client and server libraries for working with the API standard, one of which I've already used here (marshmallow-jsonapi). 

I've also added support for SampleFilter resources, because I need them for my frontend work.

TMiguelT commented:
I realised that I am using a number of features from Marshmallow version 3, which I only just found out is not compatible with Python 3.4. I'm hoping that's not a huge problem.

TMiguelT commented:
This is ready for review, by the way, @ewels.

ewels commented:
Away on vacation now sorry, but will be back at work next month!

TMiguelT commented:
Ah, apologies. Let me know when you're back!

tsnowlan commented on megaqc/model/models.py:
```diff
@@ -25,26 +24,33 @@ class Report(db.Model, CRUDMixin):
     report_id = Column(Integer, primary_key=True)
-    user_id = Column(Integer, ForeignKey('users.user_id'), index=True)
+    # If the user is deleted, we still want to retain the report
+    user_id = Column(Integer, ForeignKey('users.user_id', ondelete='SET NULL'), index=True)
```
I feel user accounts should be deactivated as opposed to deleted for historical purposes. This is good as a precaution ensure the report is retained in the event of an accidental or "accidental" deletion though.

TMiguelT replied:
That's probably reasonable. But yes, that's not really what this snippet is doing. It's just ensuring database integrity if the user *is* somehow deleted.

TMiguelT commented:
Incidentally, this PR is made mostly redundant by #98, but since that PR is based off this one, feedback on this is probably still helpful.
